marklogic-client-api/src/main/java/com/marklogic/client/datamovement/ZipSplitter.java

```diff
@@ -0,0 +1,363 @@
+     */
+    public Predicate<ZipEntry> getEntryFilter() {
+        return this.entryFilter;
+    }
+
+    /**
+     * Used to set entryFilter to splitter.
+     * @param entryFilter the filter that applied to each zipEntry
+     */
+    public void setEntryFilter(Predicate<ZipEntry> entryFilter) {
+        this.entryFilter = entryFilter;
+    }
+
+    /**
+     * Returns the uriTransformer set to splitter.
+     * @return the uriTransformer set to splitter
+     */
+    public Function<String, String> getUriTransformer() {
+        return this.uriTransformer;
+    }
+
+    /**
+     * Used to set uriTransformer to splitter
+     * @param uriTransformer the uriTransformer which applied on each document URI
+     */
+    public void setUriTransformer(Function<String, String> uriTransformer) {
+        this.uriTransformer = uriTransformer;
+    }
+
+    /**
+     * Create a new ZIP splitter.
+     */
+    public ZipSplitter() {
+        extensionFormats = new HashMap<>();
+        extensionFormats.put("", Format.UNKNOWN);
+        extensionFormats.put("json", Format.JSON);
+        extensionFormats.put("txt", Format.TEXT);
+        extensionFormats.put("xml", Format.XML);
+    }
+
+    @Override
+    public long getCount() {
+        return this.count;
+    }
+
+    /**
+     * Takes a input stream of a ZIP file and convert it to a stream of BytesHandle.
+     * The ZIP file could contain XML, JSON, TXT and BINARY files.
+     * @param input is the incoming input stream
+     * @return a stream of BytesHandle
+     * @throws IOException if input cannot be split
+     */
+    @Override
+    public Stream<BytesHandle> split(InputStream input) throws IOException {
+        if (input == null) {
+            throw new IllegalArgumentException("Input cannot be null");
+        }
+
+        if (!(input instanceof ZipInputStream)) {
+            throw new IllegalArgumentException("Input should be an instance of ZipInputStream");
+        }
+
+        return split((ZipInputStream) input);
+    }
+
+    /**
+     * Takes a ZipInputStream of a ZIP file and convert it to a stream of BytesHandle.
+     * The ZIP file could contain XML, JSON, TXT and BINARY files.
+     * @param input is a ZipInputStream of a zip file
+     * @return a stream of BytesHandle
+     * @throws IOException if input cannot be split
+     */
+    public Stream<BytesHandle> split(ZipInputStream input) throws IOException {
+        if (input == null) {
+            throw new IllegalArgumentException("Input cannot be null");
+        }
+
+        BytesHandleSpliterator bytesHandleSpliterator = new BytesHandleSpliterator(this);
+        bytesHandleSpliterator.setZipStream(input);
+        bytesHandleSpliterator.setEntryFilter(this.entryFilter);
+        bytesHandleSpliterator.setExtensionFormats(this.extensionFormats);
+        bytesHandleSpliterator.setUriTransformer(this.uriTransformer);
+
+        return StreamSupport.stream(bytesHandleSpliterator, false);
```

Comment:
Given our decision that the order of documents within the stream isn't important, maybe the second argument (whether the created stream should support parallel processing of the produced byte handles) should be true.

What do you think?

Just to be clear, the same question applies to the other call to StreamSupport.stream() in splitWriteOperations()

Reply:
I agree. I could set the second parameter to true to enable parallel processing. Then should we add Splitorator.CONCURRENT to BytesHandleSpliterator as well?

Reply:
As I read the JavaDoc for Spliterator.CONCURRENT ("Characteristic value signifying that the element source may be safely concurrently modified"), it is about concurrent modification of an item rather than concurrent processing of items in the stream.

If that reading is correct, we CONCURRENT probably isn't appropriate, but the question is entirely appropriate.

Reply:
Got it. Thank you for the answer.